python/tvm/meta_schedule/testing/validate_database.py

```diff
@@ -0,0 +1,277 @@
+    scheduled_mod: IRModule,  # compiled for "target"
+    *,
+    baseline_target: Union[str, Target],
+    target: Union[str, Target],
+    dev_type: str,
+    rpc_config: ms.runner.RPCConfig,
+    f_input_generator: Union[
+        str, Callable[[IRModule], List[tvm.nd.NDArray]]
+    ] = default_input_generator,
+    f_check_metric: Union[
+        str, Callable[[tvm.nd.NDArray, tvm.nd.NDArray], bool]
+    ] = default_check_metric,
+) -> bool:
+    """Function to validate the correctness of a scheduled module.
+
+    Parameters
+    ----------
+    original_mod : IRModule
+        The original module to be compiled.
+    scheduled_mod : IRModule
+        The scheduled module to be compiled.
+    target : Target
+        The target to compile the scheduled module.
+    rpc_config : RPCConfig
+        The RPCConfig to run the scheduled module.
+    f_input_generator : Union[str, Callable]
+        The function to generate the input data.
+    f_check_metric : Union[str, Callable]
+        The function to check the metric.
+
+    Returns
+    -------
+    result : ...
+        The result of the validation.
+    """
+
+    def to_numpy(a: List[tvm.nd.NDArray]) -> List[np.ndarray]:
+        """Convert a list of TVM NDArray to a list of numpy array"""
+        assert a is not None, "Empty result cannot be converted to numpy"
+        return [x.numpy() for x in a]
+
+    def to_tvm_ndarray(a: List[np.ndarray]) -> List[tvm.nd.NDArray]:
+        """Convert a list of numpy array to a list of TVM NDArray"""
+        assert a is not None, "Empty result cannot be converted to TVM NDArray"
+        return [tvm.nd.array(x) for x in a]
+
+    def build_and_run(mod: IRModule, target: Target, dev_type: str) -> np.ndarray:
+        """Build and run the module on the target device."""
+        rt_mod = tvm.build(mod, target=target)
+        return run_module_via_rpc(
+            rpc_config=rpc_config,
+            lib=rt_mod,
+            dev_type=dev_type,
+            args={i: v for i, v in enumerate(inputs)},  # pylint: disable=unnecessary-comprehension
+            continuation=create_computer(backend="tir"),
+            backend="tir",
+        )
+
+    # make targets
+    target = Target(target)
+    baseline_target = Target(baseline_target)
+    # fetch functions & prepare inputs
+    if isinstance(f_input_generator, str):
+        f_input_generator = get_global_func(f_input_generator)
+    if isinstance(f_check_metric, str):
+        f_check_metric = get_global_func(f_check_metric)
+    inputs = to_numpy(f_input_generator(original_mod))  # type: ignore
+    # build & run original result
+    original_res = to_numpy(build_and_run(original_mod, target=baseline_target, dev_type="cpu"))
+    scheduled_res = to_numpy(build_and_run(scheduled_mod, target=target, dev_type=dev_type))
+    # check metric
+    if f_check_metric(to_tvm_ndarray(original_res), to_tvm_ndarray(scheduled_res)):  # type: ignore
```

Comment:
To tvn ndarray actually introduces data copy. Do we need to do `to_numpy` and `to_tvm_ndarray` here? What about just pass the tvm ndarray to `f_check_metric` and only convert to numpy if needed?

Reply:
The reason we have to do the data copy is that we would like to allow usage of self-customized function as `f_check_metric` and `f_input_generator` using only their name in global registry. Therefore, the functions will only take tvm object as input so we have to convert them to tvm ndarray here. Would be great if you have other ideas to work around!

Reply:
i mean if we drop the to_numpy and directly pass original/scheduled_res to `f_check_metric`, would that work?

Reply:
Unfortunately if we keep the result as tvm nd array it will prevent the remote rpc session from closing, and the scheduled IRModule cannot be run via rpc successfully.